modules/service/src/main/scala/lucuma/odb/graphql/mapping/MutationMapping.scala

```diff
@@ -95,12 +101,34 @@ trait MutationMapping[F[_]: MonadCancelThrow]
 
   // Field definitions
 
-  private val CreateObservation =
-    MutationField("createObservation", CreateObservationInput.Binding) { (input, child)=>
-      observationService.use { svc =>
-        svc.createObservation(input.programId, input.SET.getOrElse(ObservationPropertiesInput.Default)).map(
-          _.map(id => Unique(Filter(ObservationPredicates.hasObservationId(id), child)))
-        )
+  private val CreateObservation: MutationField =
+    MutationField("createObservation", CreateObservationInput.Binding) { (input, child) =>
+
+      val createObs: F[Result[(Observation.Id, Query)]] =
+        observationService.use { svc =>
+          svc.createObservation(input.programId, input.SET.getOrElse(ObservationPropertiesInput.Default)).map(
+            _.fproduct(id => Unique(Filter(ObservationPredicates.hasObservationId(id), child)))
+          )
+        }
+
+      def insertAsterism(oid: Option[Observation.Id]): F[Result[Unit]] =
+        oid.flatTraverse { o =>
+          input.asterism.toOption.traverse { a =>
+            asterismService.use(_.insertAsterism(input.programId, o, a))
+          }
+        }.map(_.getOrElse(Result.unit))
+
+      pool.use { s =>
+        s.transaction.use { xa =>
+          for {
+            rTup  <- createObs
+            oid    = rTup.toOption.map(_._1)
+            rUnit <- insertAsterism(oid)
+            query  = (rTup, rUnit).parMapN { case ((_, query), _) => query }
+            // Fail altogether if there was an issue, say, creating the asterism
+            _     <- query.left.fold(Applicative[F].unit)(_ => xa.rollback.void)
```

Comment:
Is this correct?  I assume I need an explicit transaction and rollback here.  I want to handle an exception generated when creating the asterism and turn it into a failure `Result` and yet have it rollback the whole observation creation.

Also, there's surely a cleaner way to do `Ior.left.fold(Applicative[F].unit)(_ => xa.rollback.void)`.

Reply:
Yep, thanks!